### PR TITLE
Infer taskhost needed Fixes #6357

### LINKED ISSUE #6357: Detect architecture-specific task assemblies and infer taskhost
Today, if you have a task that is built for 32-bit or 64-bit, but does not [specify architecture in its `UsingTask`](https://docs.microsoft.com/visualstudio/msbuild/how-to-configure-targets-and-tasks?view=vs-2019#usingtask-attributes-and-task-parameters), you get an error message:

```
S:\work\cross-arch-tasks\TaskUse.proj(8,5): error MSB4062: The "Task32" task
could not be loaded from the assembly
S:\work\cross-arch-tasks\Task32\bin\Debug\net472\Task32.dll. Could not load file
or assembly
'file:///S:\work\cross-arch-tasks\Task32\bin\Debug\net472\Task32.dll' or one of
its dependencies. An attempt was made to load a program with an incorrect
format. Confirm that the <UsingTask> declaration is correct, that the assembly
and all its dependencies are available, and that the task contains a public
class that implements Microsoft.Build.Framework.ITask.
```

This isn't very actionable.

MSBuild _could_ examine the task assembly before loading it and infer an `Architecture` requirement if the assembly isn't MSIL. If we don't infer, we could at least emit a more-specific error message like

> The task assembly XXX specifies the x86 architecture but the currently-loaded .NET runtime is x64. Consider adding an `Architecture` requirement to the `UsingTask`.

### PR DESCRIPTION
**Pull request by Forgind**
Fixes #6357

### Context
If the user makes a 32-bit task and doesn't specify that in the using task and also doesn't explicitly request a task host, they get a confusing message about not being able to load some file or assembly when building with 64-bit MSBuild. Same is true for the reverse situation.

### Changes Made
Currently, it seems to correct identify MSBuild's bitness and that of a task and scrapes the information appropriately. It makes a task host that also infers that it should not load the assembly fully, which causes an issue.

### Notes
Before I investigate the issue I mentioned in Changes Made, I thought I should check the scope of this issue. Thoughts on just sending a better error message when I see that (smaller, easier change) versus pushing forward on automatically "fixing" it for the user?

### CODE REVIEW

**Comment by Forgind** [comment]
Note that this is based on #7132.

**Comment by rainersigwald** [comment]
> Before I investigate the issue I mentioned in Changes Made, I thought I should check the scope of this issue. Thoughts on just sending a better error message when I see that (smaller, easier change) versus pushing forward on automatically "fixing" it for the user?

I would like to pursue the automatic fix because it will smooth the transition, but we could have an intermediate state with a better error.

**Comment by baronfel** [comment]
There's some concern over the increase in size for the new assemblies - how much of an increase are we talking per new supported combination?  If it's small on the order of 5MB per, then that seems like a win. Otherwise we might need to look at the impacts to SDK size overall and prioritize combinations.

@rainersigwald do you have any insight here?

**Comment by rainersigwald** [comment]
@baronfel for this PR there is no size impact; it applies to .NET Framework MSBuild in VS/Build Tools scenarios which all always have 32- and 64-bit `MSBuild.exe` (.NET 4.7.2) and `MSBuildTaskHost.exe` (.NET 3.5).

Size impact is in #711 and would be on the order of 15MB/runtime environment (based on the recursive size of `C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64` in my dogfood dev17.

**Comment by baronfel** [comment]
I found the same folder and compressed it with Window's built in compression utilities to simulate what the situation would be from the SDK download perspective - it zips to 5.4 MB, which for 2 additional sets of assemblies (x64/x86) is 10.8 MB, and the current SDK for Windows is 188MB shipped, for a total increase of 5.75% increase. That seems acceptable, but I wonder - in the future could MSBuild for the target SDK platform be included in the SDK download, and the various cross-compat assemblies be shipped as an optional workloads?

File: src/Shared/TypeLoader.cs [hunk 1]
```diff
@@ -285,65 +307,370 @@ internal AssemblyInfoToLoadedTypes(Func<Type, object, bool> typeFilter, Assembly
...
+            }
+
+            private bool Required32Bit(ProcessorArchitecture arch)
+            {
+                return arch == ProcessorArchitecture.X86 || arch == ProcessorArchitecture.Arm;
+            }
+
+            private bool Required64Bit(ProcessorArchitecture arch)
+            {
+                return arch == ProcessorArchitecture.IA64 || arch == ProcessorArchitecture.Amd64;
```
Comment (akoeplinger) [inline src/Shared/TypeLoader.cs:346]: this is probably missing ARM64?

Comment (Forgind) [reply src/Shared/TypeLoader.cs:346]: For now yes, but it's still a draft 🙂

**Pull request closed without merging**